sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+    elementary
+    jordan_normal
+    invertible
+    idempotent
+
+    References
+    ==========
+    .. [1] Lang, S., Algebra, DOI 10.1007/978-1-4613-0041-0, (2002)
+    .. [2] https://en.wikipedia.org/wiki/nilpotent
+
+    """
+    def numbers_with_sum(n, k):
+        """n numbers with sum dim"""
+        if n == 1:
+            return [k]
+        num = _sample(range(k - n)) + 1
+        return [num] + numbers_with_sum(n - 1, k - num)
+
+    rank = dim - 1 if rank is None else rank
+    index = numbers_with_sum(dim - rank, dim)
+    spec = tuple((0, i) for i in index)
+    normal_form = jordan_normal(dim, spec=spec)
+    s = invertible(dim, scalars, units, length)
+    return s.inv() * normal_form * s
+
+
+def diagonalizable(dim,
+                   spec=_elementary_scalars,
+                   scalars=_elementary_scalars,
+                   units=_elementary_units,
+                   length=None):
+    r"""a square matrix n x n of a given rank
+
+    Explanation
+    ===========
+    Creates a square matrix of a given rank
+    which is diagonalizable.
+
+    The matrix will be a product
+
+    .. math::
+
+        \mathbf{S} \cdot \mathbf{D} \cdot \mathbf{S}^{-1}
+
+    of an :func:`invertible` matrix $\mathbf{S}$ and
+    a :func:`diagonal_normal` matrix of given rank
+    and eigenvalues (diagonal entries) from **spec**.
+
+    To obtain a matrix with *complex* *eigenvalues*
+    but non-complex entries,
+    e.g. $\lambda_{1/2}=a \pm b \ i$,
+    you better may use a product of an invertible matrix $\mathbf{S}$
+    with non-complex entries (**scalars**)
+    and where $\mathbf{D}$ is a non-complex block diagonal matrix
+    with a block $\mathbf{D}_z$
+
+    .. math::
+
+        \mathbf{D}_z = \left[\begin{array}{cc}a & b \\ -b & a\end{array}\right]
+
+    Examples
+    ========
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import diagonalizable
+    >>> seed(1)
+
+    >>> diagonalizable(3)
+    Matrix([
+    [-1,  0, 0],
+    [-2, -3, 4],
+    [-2, -2, 3]])
+
+    >>> m = diagonalizable(3, spec=(1,2,2), scalars=(1,2,3), length=10)
+    >>> m
+    Matrix([
+    [ 4,  2, -6],
+    [-6, -4, 18],
+    [-1, -1,  5]])
+    >>> m.eigenvals(multiple=True)
+    [1, 2, 2]
+    >>> m.jordan_form(calc_transform=False)
+    Matrix([
+    [1, 0, 0],
+    [0, 2, 0],
+    [0, 0, 2]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension of matrix
+    spec : tuple or list of symbols (optional)
+        set of values of which scalars (diagonal entries) are choosen
+        see :func:`diagonal_normal`
+    scalars : tuple or list of symbols (optional)
+        see :func:`invertible`
+    units : tuple or list of symbols (optional)
+        see :func:`invertible`
+    length : integer (optional with default 2 * **dim**)
+        see :func:`invertible`
+
+    """
+
+    normal_form = diagonal_normal(dim, spec)
+    s = invertible(dim, scalars, units, length)
+    return s.inv() * normal_form * s
```

Comment:
Does the method used by `invertible` make it possible to compute the inverse of `s` more efficiently than just computing `s` and calling `s.inv()`?

Reply:
What I'm wondering is if it could be more efficient to generate both a matrix and its inverse simultaneously like:
```python
sinv, s = invertible_with_inverse(...)
return sinv * normal_form * s
```
The code that builds `s` has access to all the matrices that make up `s` which are probably more easily invertible.

Reply:
This might work but the matrices to build `s` are quite low level (in `super_elementary`). So feeding them up the chain ist quite a bit of work. I'll take it on my backlog for a future version.

Reply:
It seems that most cases where `invertible` is called by the other functions here also want the inverse.

Reply:
Calculating the inverse can be a lot more expensive for matrices with different kinds of entries rather than just integers.